Functions/xASL_quant_Basil.m

```diff
@@ -123,170 +130,193 @@
-
-    % Labelling type - PASL or pCASL
-    switch lower(x.Q.LabelingType)
-        case 'pasl'
-            fprintf('Basil: PASL model\n');
-        case 'casl'
-            fprintf(FIDoptionFile, '--casl\n');
-            fprintf('Basil: CASL/PcASL model\n');
-    end
-
-
-    %% Model option - 1 or 2 compartment
-    % The 1-compartment model is what we call 'White paper mode' in oxford-asl.
-    % This means zero ATT (all bolus delivered by imaging time) and blood T1 only
-    % The 2-compartment model is the 'standard' buxton model which takes into account
-    % the ATT and tissue T1 value.
-
-    switch x.Q.nCompartments
-        case 1
-            fprintf(FIDoptionFile, '--bat=0\n');
-            fprintf(FIDoptionFile, '--t1=%f\n', x.Q.BloodT1/1000);
-            fprintf('Basil: Single-compartment (white paper mode)\n');
-        case 2
-            fprintf(FIDoptionFile, '--bat=%f\n', x.Q.ATT/1000);
-            fprintf(FIDoptionFile, '--t1=%f\n', x.Q.TissueT1/1000);
-            fprintf('Basil: 2-compartment - ATT=%fs\n', x.Q.ATT/1000);
-    end
+%% Create option_file that contains options which are passed to Fabber
+% basil_options is a character array containing CLI args for the Basil command
+
+FIDoptionFile = fopen(pathBasilOptions, 'w+');
+BasilOptions = '';
+
+fprintf(FIDoptionFile, '# Basil options written by ExploreASL\n');
+fprintf(FIDoptionFile, '--iaf=diff\n'); % as input is PWI
+
+%% Basic tissue parameters
+fprintf(FIDoptionFile, '--t1b=%f\n', x.Q.BloodT1/1000);
+fprintf(FIDoptionFile, '--t1=%f\n', x.Q.TissueT1/1000);
+
+%% Basic acquisition parameters
+
+% Labelling type - PASL or pCASL
+switch lower(x.Q.LabelingType)
+	case 'pasl'
+		fprintf('Basil: PASL model\n');
+		TIs = (unique(x.Q.Initial_PLD))'/1000;
+
+		% Print all the TIs
+		if x.modules.asl.bTimeEncoded || x.modules.asl.bMultiPLD
+			% For Time-encoded, we skip the first volume
+			if x.modules.asl.bTimeEncoded
+				TIs = TIs(2:end);
+			end
+			for iTI = 1:length(TIs)
+				fprintf(FIDoptionFile, '--ti%d=%.2f\n', iTI, TIs(iTI));
+			end
+			
+		else
+			fprintf(FIDoptionFile, '--ti=%.2f\n', TIs);
+		end
+		
+		% Either print bolus duration or unspecify it
+		if isfield(x.Q,'LabelingDuration') && x.Q.LabelingDuration
+			fprintf(FIDoptionFile, '--tau=%.2f\n', x.Q.LabelingDuration/1000);
+		else
+			% Bolus duration not know. If multi-TI, then try to infer it
+			if length(TIs) > 1
+				fprintf(FIDoptionFile, '--infertau\n');
+				fprintf('Basil: Infer bolus duration component\n')
+			end
+		end
+		
+	case {'casl','pcasl'}
+		fprintf(FIDoptionFile, '--casl\n');
+		fprintf('Basil: CASL/PCASL model\n');
+		% Print all the PLDs and LabDurs
+		
+		[PLDs, ind] = unique(x.Q.Initial_PLD);
+		PLDs = PLDs'/1000;
+		if length(x.Q.LabelingDuration)>1
+			LDs = (x.Q.LabelingDuration(ind))'/1000;
+		else
+			LDs = ones(size(PLDs))*x.Q.LabelingDuration/1000;
+		end
+		% For Time-encoded, we skip the first volume
+		if x.modules.asl.bTimeEncoded
+			PLDs = PLDs(2:end);
```

Comment:
We have this twice - for PASL and PCASL. And for PCASL, we need to do this for LD as well. So doing it at one location for PASL and PCASL with PLD would mess it up for LD. So I would keep this. Unless you see it more than twice.

Reply:
OK perhaps just nice to avoid 1 extra PLD or LD variable and do this at a single location. Is easier to read for others, you can then do this at the top of the function with the admin, so you don't have to through the function.

Reply:
Agreed that it is more transparent to keep as it is. Can you resolve this?